userspace/async/async_metadata_source.tpp

```diff
@@ -0,0 +1,286 @@
+limitations under the License.
+
+*/
+#include "logger.h"
+
+#include <assert.h>
+#include <algorithm>
+#include <chrono>
+#include <iostream>
+#include <list>
+#include <string>
+
+namespace sysdig
+{
+
+template<typename key_type, typename metadata_type>
+async_metadata_source<key_type, metadata_type>::async_metadata_source(
+		const uint64_t max_wait_ms,
+		const uint64_t ttl_ms):
+	m_max_wait_ms(max_wait_ms),
+	m_ttl_ms(ttl_ms),
+	m_thread(),
+	m_running(false),
+	m_terminate(false),
+	m_mutex(),
+	m_queue_not_empty_condition(),
+	m_metadata_map()
+{ }
+
+template<typename key_type, typename metadata_type>
+async_metadata_source<key_type, metadata_type>::~async_metadata_source()
+{
+	try
+	{
+		stop();
+	}
+	catch(...)
+	{
+		g_logger.log(std::string(__FUNCTION__) +
+		             ": Exception in destructor",
+		             sinsp_logger::SEV_ERROR);
+	}
+}
+
+template<typename key_type, typename metadata_type>
+uint64_t async_metadata_source<key_type, metadata_type>::get_max_wait() const
+{
+	return m_max_wait_ms;
+}
+
+template<typename key_type, typename metadata_type>
+uint64_t async_metadata_source<key_type, metadata_type>::get_ttl() const
+{
+	return m_ttl_ms;
+}
+
+template<typename key_type, typename metadata_type>
+void async_metadata_source<key_type, metadata_type>::stop()
+{
+	bool join_needed = false;
+
+	{
+		std::unique_lock<std::mutex> guard(m_mutex);
+
+		if(m_running)
+		{
+			m_terminate = true;
+			join_needed = true;
+
+			// The async thread might be waiting for new events
+			// so wake it up
+			m_queue_not_empty_condition.notify_one();
+		}
+	} // Drop the mutex before join()
+
+	if (join_needed)
+	{
+		m_thread.join();
+
+		// Remove any pointers from the thread to this object
+		// (just to be safe)
+		m_thread = std::thread();
+	}
+}
+
+template<typename key_type, typename metadata_type>
+bool async_metadata_source<key_type, metadata_type>::is_running() const
+{
+	std::lock_guard<std::mutex> guard(m_mutex);
+
+	return m_running;
+}
+
+template<typename key_type, typename metadata_type>
+void async_metadata_source<key_type, metadata_type>::run()
+{
+	m_running = true;
+
+	while(!m_terminate)
+	{
+		{
+			std::unique_lock<std::mutex> guard(m_mutex);
+
+			while(!m_terminate && m_request_queue.empty())
+			{
+				// Wait for something to show up on the queue
+				m_queue_not_empty_condition.wait(guard);
+			}
+
+			prune_stale_requests();
+		}
+
+		if(!m_terminate)
+		{
+			run_impl();
+		}
+	}
+
+	m_running = false;
+}
+
+template<typename key_type, typename metadata_type>
+bool async_metadata_source<key_type, metadata_type>::lookup(
+		const key_type& key,
+		metadata_type& metadata,
+		const callback_handler& callback)
+{
+	std::unique_lock<std::mutex> guard(m_mutex);
+
+	if(!m_running)
+	{
+		m_thread = std::thread(&async_metadata_source::run, this);
+	}
+
+	typename metadata_map::const_iterator itr = m_metadata_map.find(key);
+	bool request_complete = (itr != m_metadata_map.end()) && itr->second.m_available;
```

Comment:
some questions on the semantics:
1) if we have  atimeout set, should we not be trying to fetch the new data for at least timeout s before falling back to the cached data?
2) if we have the data, should we not fetch again? Since that's going to cause us to oscillate between having data and not having data

Reply:
The semantics are you call `lookup()`, then we wait as long as you're willing.  If we don't get an answer within that time, we let you know and we keep working on getting what you asked for.  When we eventually get it, we save it for you and on the next call to `lookup()` we give you that value.  We explicitly do not want to re-query in this case.

In the metadata use case --- the case we're focused on here --- we don't expect that the data changes, so "new" data is as good as cached data.